File: src/main/java/com/pubnub/api/PubNubException.java
package com.pubnub.api;

import com.google.gson.JsonElement;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.ToString;
import retrofit2.Call;

@Builder
@Getter
@ToString
public class PubNubException extends Exception {
    private String errormsg;
    private PubNubError pubnubError;
    private JsonElement jso;
    private String response;
    private int statusCode;

    @Getter(AccessLevel.NONE)
    @ToString.Exclude
    private Call affectedCall;
}

File: pubnub-java/src/main/kotlin/com/pubnub/api/crypto/exception/PubNubException.kt
package com.pubnub.api.crypto.exception

import retrofit2.Call

/**
 * Custom exception wrapper for errors occurred during execution or processing of a PubNub API operation.
 *
 * @property errorMessage The error message received from the server, if any.
 * @property pubnubError The appropriate matching PubNub error.
 * @property jso The error json received from the server, if any.
 * @property statusCode HTTP status code.
 * @property affectedCall A reference to the affected call. Useful for calling [retry][Endpoint.retry].
 */
data class PubNubException(
    val errorMessage: String? = null,
    val pubnubError: PubNubError? = null,
    val jso: String? = null,
    val statusCode: Int = 0,
    val affectedCall: Call<*>? = null
) : Exception(errorMessage) {

    internal constructor(pubnubError: PubNubError) : this(
        errorMessage = pubnubError.message,
        pubnubError = pubnubError
    )
}
